Introduce couch_file:commit_header/2

iilyak opened the pull request:
## Overview

This PR is an optimization to reduce number of messages to couch_file proccess, in casses when we need to call sync. The couch_file:commit_header/2 is more efficient equivalent of:

```
couch_file:sync(Fd),
ok = couch_file:write_header(Fd, NewHeader),
couch_file:sync(Fd),
```

## Testing recommendations

Existing testing suite should suffice. 

## Related Issues or Pull Requests

<!-- If your changes affect multiple components in different
     repositories please put links to those issues or pull requests here.  -->

## Checklist

- [x] Code is written and works correctly
- [x] Changes are covered by tests
- [ ] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [ ] Documentation changes were made in the `src/docs` folder
- [ ] Documentation changes were backported (separated PR) to affected branches

iilyak committed:
Introduce couch_file:commit_header/2

This PR is an optimization to reduce number of messages to couch_file proccess,
in casses when we need to call sync. The couch_file:commit_header/2 is more
efficient equivalent of:

```
couch_file:sync(Fd),
ok = couch_file:write_header(Fd, NewHeader),
couch_file:sync(Fd),
```

rnewson commented on src/couch/src/couch_file.erl:
```diff
@@ -596,7 +604,28 @@ handle_call({write_header, Bin}, _From, #file{fd = Fd, eof = Pos} = File) ->
-    {reply, find_header(Fd, Pos div ?SIZE_BLOCK), File}.
+    {reply, find_header(Fd, Pos div ?SIZE_BLOCK), File};
+handle_call({commit_header, Bin}, From, #file{fd = Fd} = File) ->
+    case fsync(Fd) of
```
a `maybe` expression would be clearer and is available in all versions of OTP that couchdb supports.

iilyak replied:
rewritten here c4d538a232130b11366fe4ede459100d11ba5d2b

rnewson commented on src/couch/src/couch_file.erl:
```diff
@@ -596,7 +604,28 @@ handle_call({write_header, Bin}, _From, #file{fd = Fd, eof = Pos} = File) ->
+handle_call({commit_header, Bin}, From, #file{fd = Fd} = File) ->
+    case fsync(Fd) of
+        ok ->
+            Result = handle_call({write_header, Bin}, From, File),
```
I find this quite ugly, can we extract the common code into a function that both places call please?

iilyak replied:
addressed in 050ae0742a10

iilyak committed:
Revert "Introduce couch_file:commit_header/2"

This reverts commit 62d78c7630b72630f3d999fe235e23565401373e.

iilyak committed:
Factor out the do_write_header

iilyak committed:
Introduce couch_file:commit_header/2

This PR is an optimization to reduce number of messages to couch_file proccess,
in casses when we need to call sync. The couch_file:commit_header/2 is more
efficient equivalent of:

```
couch_file:sync(Fd),
ok = couch_file:write_header(Fd, NewHeader),
couch_file:sync(Fd),
```

nickva reviewed (changes requested):
Not a bad idea but there is a fair amount of duplicated code and the inline handle_call() call is a not as clean-looking, as @rnewson mentioned.

Maybe instead of a separate command/call altogether, add an option to the write_header call to sync or not, defaulting to not not sync. Sometimes boolean flags are not as descriptive, so we can make it into an options list `[sync]`? 

Using of maybe sounds intriguing, good idea, @rnewson!

nickva commented on src/couch/src/couch_bt_engine_compactor.erl:
```diff
@@ -660,8 +660,7 @@ commit_compaction_data(#st{header = OldHeader} = St0, Fd) ->
     },
-    ok = couch_file:sync(Fd),
-    ok = couch_file:write_header(Fd, CompHeader),
+    ok = couch_file:commit_header(Fd, CompHeader),
```
It seems here we're changing the behavior and adding an extra sync after we write the header. We only commit periodically in the compactor, and can't think this would be an issue at first sight but thought it's worth pointing it out at least.

In other words, at least in the compactor it's not necessarily a pure perf optimization as fsync calls be a bit more expensive, especially with remote block storage involved.

iilyak commented:
> Maybe instead of a separate command/call altogether, add an option to the write_header call to sync or not, defaulting to not not sync. Sometimes boolean flags are not as descriptive, so we can make it into an options list `[sync]`?

Handling flag doesn't land naturally into maybe pattern. The `maybe` version (without flag) would look like:

```erlang
handle_call({commit_header, Bin}, _From, #file{fd = Fd} = File) ->
    maybe
      ok ?= fsync(Fd),
      {ok, NewFile} ?= do_write_header(Bin, File),
      ok ?= fsync(Fd),
      {reply, ok, NewFile}
    else
      {{error, _} = Error, State} ->
        {reply, Error, State};
      {error, _} = Error ->
        % We're intentionally dropping all knowledge
        % of this Fd so that we don't accidentally
        % recover in some whacky edge case that I
        % can't fathom.
        {stop, Error, Error, #file{fd = nil}}
    end.
```

The `maybe` matches do not support guards and also the `?=` is only at the top level of `maybe/end`.

iilyak committed:
Use commit_header/2 in couch_bt_engine:commit_data/1

rnewson commented:
hm agree, the maybe version is not straightforward, which would really be the point of doing it. sorry for the bad steer.

I quite like the idea of an extra argument and an upgrade clause (defaulting to `false` or `[]` or whatever we go with there) and then call it with `true` or `[sync]`

rnewson commented:
at this point I'm going to butt out because this is classic bikeshedding on my part.

iilyak committed:
fixup! Factor out the do_write_header

iilyak committed:
fixup! Introduce couch_file:commit_header/2
